File: frontend/lib/core/models/support_message.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'support_message.freezed.dart';
part 'support_message.g.dart';

/// Конвертер для преобразования ID из int (PostgreSQL) или String (UUID) в String
String _supportMessageIdToString(dynamic value) {
  return value.toString();
}

/// Конвертер для безопасного парсинга даты из JSON
DateTime _dateTimeFromJson(dynamic value) {
  if (value == null) {
    throw ArgumentError('Дата не может быть null');
  }
  if (value is DateTime) {
    return value;
  }
  if (value is String) {
    try {
      return DateTime.parse(value);
    } catch (e) {
      print('[SupportMessage] Ошибка парсинга даты: $value, ошибка: $e');
      return DateTime.now();
    }
  }
  if (value is int) {
    return DateTime.fromMillisecondsSinceEpoch(value * 1000);
  }
  print('[SupportMessage] Неожиданный тип для даты: ${value.runtimeType}, значение: $value');
  return DateTime.now();
}

@freezed
class SupportMessage with _$SupportMessage {
  const factory SupportMessage({
    @JsonKey(fromJson: _supportMessageIdToString) required String id,
    @JsonKey(name: 'user_id') String? userId,
    String? name,
    String? email,
    required String type, // 'suggestion', 'bug', 'question'
    required String message,
    required String status, // 'new', 'answered', 'closed'
    @JsonKey(name: 'created_at', fromJson: _dateTimeFromJson) required DateTime createdAt,
    @JsonKey(name: 'updated_at', fromJson: _dateTimeFromJson) required DateTime updatedAt,
    @Default(false) @JsonKey(name: 'has_unread_replies') bool hasUnreadReplies,
    @Default(0) @JsonKey(name: 'replies_count') int repliesCount,
  }) = _SupportMessage;

  factory SupportMessage.fromJson(Map<String, dynamic> json) => _$SupportMessageFromJson(json);
}

File: backend/app/models/support_message.py
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Boolean
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from ..db import Base


class SupportMessage(Base):
    """Модель для сообщений пользователей в поддержку"""
    __tablename__ = "support_messages"

    id = Column(UUID(as_uuid=True), primary_key=True, index=True, server_default=func.gen_random_uuid())
    user_id = Column(String, nullable=False, index=True)  # UUID пользователя как строка
    name = Column(String(255), nullable=False)
    email = Column(String(255), nullable=False)
    type = Column(String(50), nullable=False)  # 'suggestion', 'bug', 'question'
    message = Column(Text, nullable=False)
    status = Column(String(50), nullable=False, default="new", index=True)  # 'new', 'answered', 'closed'
    
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    replies = relationship("SupportMessageReply", back_populates="support_message", cascade="all, delete-orphan")


class SupportMessageReply(Base):
    """Модель для ответов на сообщения поддержки"""
    __tablename__ = "support_message_replies"

    id = Column(UUID(as_uuid=True), primary_key=True, index=True, server_default=func.gen_random_uuid())
    message_id = Column(UUID(as_uuid=True), ForeignKey("support_messages.id", ondelete="CASCADE"), nullable=False, index=True)
    reply_text = Column(Text, nullable=False)
    replied_by = Column(String(255), nullable=True)  # 'telegram', 'admin_user_id', 'user_{user_id}'
    is_read = Column(Boolean, nullable=False, default=False, index=True)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    
    # Relationships
    support_message = relationship("SupportMessage", back_populates="replies")

